### ISSUE TITLE
Changing organisation name via API does not update Solr, breaks pages

### ISSUE THREAD

**Issue by ThrawnCA** [ckan/ckan#5842]
**CKAN version**

2.8.4

**Describe the bug**

Updating an organisation name via the `organization_update` API does not updated the associated Solr records for the organisation's datasets, which in turn causes errors (in `snippets/organization.html`) when viewing those dataset pages.

**Steps to reproduce**

- Create an organisation called "foo".
- Create a dataset in the organisation named "baz".
- Use the `organization_update` API to alter the organisation name to "meh".
- Visit `/dataset/baz`

**Expected behavior**
The dataset should be displayed with the new organisation name.

**Additional details**

Running a Solr index rebuild on the affected dataset(s) resolves the error.

**Comment by ThrawnCA** [comment]
Triggering an index update for every affected dataset could be quite a bit of work, of course, if an organisation contains many datasets. A cheaper alternative would be to update `snippets/organization.html` so it gracefully handles the renaming. Currently, it seems to be looking up the organisation associated with a package by using the organisation name, not UUID - so when the name changes, it breaks.

**Comment by ThrawnCA** [comment]
It appears that this does not always affect every dataset in the organisation; some will work while others will break. I haven't yet identified a pattern, although private datasets seem more likely to be affected (but public datasets _can_ break too).

**Comment by duttonw** [comment]
I have replicated this on https://demo.ckan.org/ following the steps above

1. original org https://demo.ckan.org/organization/abcdd
2. dataset createdhttps://demo.ckan.org/dataset/tadsf
3. org changed to 123abc including url https://demo.ckan.org/organization/123abc
4. https://demo.ckan.org/dataset/tadsf is now broken

```
Server Error
An internal server error occurred
```

**Comment by amercader** [comment]
This is a long standing issue. Basically if you update an organization's name you need to reindex all of its dataset so the Solr index gets the new value.
As this is potentially a long task depending on the number of datasets it should be implemented as a [background job](https://docs.ckan.org/en/2.9/maintaining/background-tasks.html).

Here's a quick spec in case someone wants to contribute it:

* Create a new `ckan/jobs.py` module (we still don't have builtin jobs), with a function named `update_organization_name_on_datasets` or similar
* This functions takes an organization id as parameter
* The function calls `package_search` with a `fq`: `owner_org:<organization_id>` filter to get all datasets that belong to the org, and then calls `ckan.lib.search.rebuild()` with their ids. Things to check: results pagination, job timeout, etc
* Add a small check at the end of `organization_update` to see if the name changed as part of the update, if so call enqueue the function above passing the org id as parameter
*

**Comment by ThrawnCA** [comment]
Would it be feasible to fix the more immediate issue by having the package linked to the organisation using its UUID rather than its name?

**Comment by amercader** [comment]
@ThrawnCA sure, if a snippet uses the org name it could use the UUID

**Comment by wardi** [comment]
@ThrawnCA yes but the search results page will need to query the DB for each result displayed which adds quite a few round-trips instead of having all the information needed returned from solr in one query.

**Comment by amercader** [comment]
This is currently being worked on by @alexkreidler
